trajopt/src/collision_terms.cpp

```diff
@@ -265,6 +267,358 @@ void CollisionsToDistanceExpressions(sco::AffExprVector& exprs,
+        tesseract_kinematics::jacobianChangeRefPoint(
+            jac, (link_transform * itB->transform.inverse()).linear() * (itB->transform * res.nearest_points_local[1]));
+
+        // Eigen::Isometry3d test_link_transform, temp1, temp2;
+        // manip->calcFwdKin(test_link_transform, dofvals, itB->link_name);
+        // temp1 = world_to_base * test_link_transform;
+        // temp2 = link_transform * itB->transform.inverse();
+        // assert(temp1.isApprox(temp2, 0.0001));
+
+        // Eigen::MatrixXd jac_test;
+        // jac_test.resize(6, manip->numJoints());
+        // tesseract_kinematics::numericalJacobian(jac_test, world_to_base, *manip, dofvals, itB->link_name,
+        // res.nearest_points_local[1]); bool check = jac.isApprox(jac_test, 1e-3); assert(check == true);
+
+        if (res.distance < worst_dist)
+          worst_dist = res.distance;
+
+        double weight = sco::pospart(data[0] - res.distance);
+        total_weight_b += weight;
+
+        dist_grad_b += (weight * scale * res.normal.transpose() * jac.topRows(3));
+      }
+    }
+
+    exprs_data.push_back(data);
+    if (!found_a && !found_b)
+    {
+      exprs.push_back(sco::AffExpr(0));
+    }
+    else
+    {
+      sco::AffExpr dist(worst_dist);
+      if (found_a)
+      {
+        dist_grad_a *= (1.0 / total_weight_a);
+        sco::exprInc(dist, sco::varDot(dist_grad_a, vars));
+        sco::exprInc(dist, -dist_grad_a.dot(dofvals));
+      }
+
+      if (found_b)
+      {
+        dist_grad_b *= (1.0 / total_weight_b);
+        sco::exprInc(dist, sco::varDot(dist_grad_b, vars));
+        sco::exprInc(dist, -dist_grad_b.dot(dofvals));
+      }
+
+      exprs.push_back(dist);
+    }
+  }
+}
+
+void CollisionEvaluator::CollisionsToDistanceExpressionsContinuousW(
+    sco::AffExprVector& exprs,
+    AlignedVector<Eigen::Vector2d>& exprs_data,
+    const tesseract_collision::ContactResultMap& dist_results,
+    const sco::VarVector& vars0,
+    const sco::VarVector& vars1,
+    const DblVec& x,
+    bool isTimestep1)
+{
+  Eigen::VectorXd dofvals0 = sco::getVec(x, vars0);
+  Eigen::VectorXd dofvals1 = sco::getVec(x, vars1);
+  Eigen::VectorXd dofvalst = Eigen::VectorXd::Zero(dofvals0.size());
+
+  // All collision data is in world corrdinate system. This provides the
+  // transfrom for converting data between world frame and manipulator
+  // frame.
+
+  exprs.clear();
+  exprs_data.clear();
+  exprs.reserve(dist_results.size());
+  exprs_data.reserve(dist_results.size());
+  for (const auto& pair : dist_results)
+  {
+    double worst_dist{ std::numeric_limits<double>::max() };
+    double total_weight_a{ 0 }, total_weight_b{ 0 };
+    bool found_a{ false }, found_b{ false };
+
+    Eigen::VectorXd dist_grad_a = Eigen::VectorXd::Zero(manip_->numJoints());
+    Eigen::VectorXd dist_grad_b = Eigen::VectorXd::Zero(manip_->numJoints());
+
+    const Eigen::Vector2d& data = safety_margin_data_->getPairSafetyMarginData(pair.first.first, pair.first.second);
```

Comment:
A more descriptive variable name might be useful here. I had to look back to figure out what data was.

Reply:
I added comments everywhere this is created.